ext/KrylovKitChainRulesCoreExt/utilities.jl

```diff
@@ -0,0 +1,64 @@
+safe_inv(a, tol) = abs(a) < tol ? zero(a) : inv(a)
+
+# vecs are assumed orthonormal
+function orthogonalprojector(vecs, n)
+    function projector(w)
+        w′ = zerovector(w)
+        @inbounds for i in 1:n
+            w′ = VectorInterface.add!!(w′, vecs[i], inner(vecs[i], w))
+        end
+        return w′
+    end
+    return projector
+end
+function orthogonalcomplementprojector(vecs, n)
+    function projector(w)
+        w′ = scale(w, 1)
+        @inbounds for i in 1:n
+            w′ = VectorInterface.add!!(w′, vecs[i], -inner(vecs[i], w))
+        end
+        return w′
+    end
+    return projector
+end
+# vecs are not assumed orthonormal, G is the Cholesky factorisation of the overlap matrix
+function orthogonalprojector(vecs, n, G::Cholesky)
+    overlaps = zeros(eltype(G), n)
+    function projector(w)
+        @inbounds for i in 1:n
+            overlaps[i] = inner(vecs[i], w)
+        end
+        overlaps = ldiv!(G, overlaps)
+        w′ = zerovector(w)
+        @inbounds for i in 1:n
+            w′ = VectorInterface.add!!(w′, vecs[i], +overlaps[i])
+        end
+        return w′
+    end
+    return projector
+end
+function orthogonalcomplementprojector(vecs, n, G::Cholesky)
+    overlaps = zeros(eltype(G), n)
+    function projector(w)
+        @inbounds for i in 1:n
+            overlaps[i] = inner(vecs[i], w)
+        end
+        overlaps = ldiv!(G, overlaps)
+        w′ = scale(w, 1)
+        @inbounds for i in 1:n
+            w′ = VectorInterface.add!!(w′, vecs[i], -overlaps[i])
+        end
+        return w′
+    end
+    return projector
+end
+
+function _realview(v::AbstractVector{Complex{T}}) where {T}
+    v_real = reinterpret(T, v)
+    return view(v_real, axes(v_real, 1)[begin:2:end])
+end
+
+function _imagview(v::AbstractVector{Complex{T}}) where {T}
+    v_real = reinterpret(T, v)
+    return view(v_real, axes(v_real, 1)[(begin + 1):2:end])
+end
```

Comment:
we might want to explicitly restrict these definitions to BlasFloats, considering it relies heavily on how the data is stored? I am not sure how the semantics of `reinterpret` are defined for eg `Complex{BigFloat}`, and I guess that KrylovKit currently would not work with these types anyways, but it might be safer to just explicitly restrict and if necessary extend afterwards?

Reply:
I think its up to reinterpret to handle the more complicated types correctly; the semantics of `reinterpret` are used correctly. I think `reinterpret` is only defined for isbits types, and will thus error for anything else.

Reply:
There might be user defined `number` types for which this is correct, which we would exclude with the restriction (not that we currently support this in the rest of KrylovKit, but you never know).